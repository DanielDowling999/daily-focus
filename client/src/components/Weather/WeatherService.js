const API_KEY = "2d6026fb2c6a6ef3bbbc1f81c56baf04";

function getWeatherData(
    lat,
    lon,
    setIcon,
    setMain,
    setTemp,
    setMaxTemp,
    setMinTemp,
    setDailyData,
    setHourlyData,
    setIsError
) {
    // Get all the weather data (current, hourly for current day, and daily)
    fetch(
        `https://api.openweathermap.org/data/2.5/onecall?lat=${lat}&lon=${lon}&units=metric&exclude=alerts,minutely&appid=${API_KEY}`
    )
        .then((res) => res.json())
        .then((response) => {
            // Current weather
            setIsError(false);
            setMain(response.current.weather[0].main);
            setTemp(Math.floor(response.current.temp));
            setMaxTemp(Math.floor(response.daily[0].temp.max));
            setMinTemp(Math.floor(response.daily[0].temp.min));
            setIcon(response.current.weather[0].icon);

            // Hourly weather for next 6 hours
            const hourlyData = response.hourly.slice(0, 6);
            setHourlyData(hourlyData);

            // Daily weather for next 5 days
            const dailyData = response.daily.slice(1, 6);
            setDailyData(dailyData);
            return response;
        })
        .catch(() => {
            setIsError(true);
        });
}

function getLocationData(lat, lon, setCity, setCountry, setIsError) {
    // Get the city and country name for the requested coordinates
    fetch(`http://api.openweathermap.org/geo/1.0/reverse?lat=${lat}&lon=${lon}&appid=${API_KEY}`)
        .then((res) => res.json())
        .then((response) => {
            setIsError(false);
            setCity(response[0].name);
            setCountry(response[0].country);
            return response;
        })
        .catch(() => {
            setIsError(true);
        });
}

export { getWeatherData, getLocationData };
