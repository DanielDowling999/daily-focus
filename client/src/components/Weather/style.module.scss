.weatherWidget {
    width: 450px;
    height: 560px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.container {
    width: 340px;
    height: 400px;
    background-color: #ffffff;
    box-shadow: 0px 4px 16px 8px rgba(0, 0, 0, 0.25);
    border-radius: 25px;
}

.headerRectangle {
    width: 340px;
    height: 47.88px;
    background: #30a0f5;
    border-radius: 25px 25px 0px 0px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.weatherTextBox {
    width: 251px;
    height: 47.88px;
    font-family: Ropa Sans;
    font-style: normal;
    font-weight: normal;
    font-size: 30px;
    display: flex;
    justify-content: center;
    align-items: center;
    color: #ffffff;
}

.currentWeatherBox {
    width: 300px;
    height: 80.8px;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
}

.temp {
    font-family: Ropa Sans;
    font-style: normal;
    font-weight: normal;
    font-size: 30px;
    display: flex;
    align-items: center;
    text-align: center;
    color: #000000;
}

.tempBounds {
    width: 82px;
    height: 20.95px;
    font-family: Ropa Sans;
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 15px;
    display: flex;
    align-items: center;
    text-align: right;

    color: #000000;
}

.weatherNZ {
    width: 179px;
    height: 19.95px;
    font-family: Ropa Sans;
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 15px;
    display: flex;
    align-items: center;
    text-align: right;
    color: #000000;
}

.iconContainer {
    font-size: 70px;
    width: 100px;
    height: 80.8px;
    float: left;
}

.hourlyForecast {
    width: 300px;
    height: 67.83px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.dailyForecast {
    width: 291px;
    height: 123.69px;
    display: flex;
    flex-direction: column;
}

.content {
    display: flex;
    height: 330px;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
}

.errorContainer {
    height: 330px;
    display: flex;
    justify-content: center;
    align-items: center;
}
