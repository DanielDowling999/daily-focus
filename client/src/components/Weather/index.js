import React, { useState, useEffect } from "react";
import DayCard from "./DayCard";
import HourCard from "./HourCard";
import Icon from "./Icon";
import { getWeatherData, getLocationData } from "./WeatherService";
import styles from "./style.module.scss";

const AUCKLAND_COORDS = { lat: -36.85, lon: 174.76 };

function Weather() {
    const [icon, setIcon] = useState(null);
    const [main, setMain] = useState(null);
    const [temp, setTemp] = useState(null);
    const [maxTemp, setMaxTemp] = useState(null);
    const [minTemp, setMinTemp] = useState(null);
    const [dailyData, setDailyData] = useState([]);
    const [hourlyData, setHourlyData] = useState([]);
    const [city, setCity] = useState(null);
    const [country, setCountry] = useState(null);
    const [isError, setIsError] = useState(false);

    const formatDayCards = (dailyData) => {
        return dailyData.map((day, index) => <DayCard day={day} key={index} />);
    };

    const formatHourCards = (hourlyData) => {
        let hourArray = hourlyData.map((hour, index) => <HourCard hour={hour} key={index} />);
        return hourArray;
    };

    useEffect(() => {
        if ("geolocation" in navigator) {
            navigator.geolocation.getCurrentPosition(
                function (position) {
                    const lat = position.coords.latitude;
                    const lon = position.coords.longitude;
                    getWeatherData(
                        lat,
                        lon,
                        setIcon,
                        setMain,
                        setTemp,
                        setMaxTemp,
                        setMinTemp,
                        setDailyData,
                        setHourlyData,
                        setIsError
                    );
                    getLocationData(lat, lon, setCity, setCountry, setIsError);
                },
                // Permission not given to access location, so default to Auckland coordinates
                function onError() {
                    getWeatherData(
                        AUCKLAND_COORDS.lat,
                        AUCKLAND_COORDS.lon,
                        setIcon,
                        setMain,
                        setTemp,
                        setMaxTemp,
                        setMinTemp,
                        setDailyData,
                        setHourlyData,
                        setIsError
                    );
                    getLocationData(
                        AUCKLAND_COORDS.lat,
                        AUCKLAND_COORDS.lon,
                        setCity,
                        setCountry,
                        setIsError
                    );
                }
            );
        } else {
            getWeatherData(
                AUCKLAND_COORDS.lat,
                AUCKLAND_COORDS.lon,
                setIcon,
                setMain,
                setTemp,
                setMaxTemp,
                setMinTemp,
                setDailyData,
                setHourlyData,
                setIsError
            );
            getLocationData(AUCKLAND_COORDS.lat, AUCKLAND_COORDS.lon, setCity, setCountry, isError);
        }
    }, []);

    return (
        <div className={styles.weatherWidget}>
            <div className={styles.container}>
                <div className={styles.headerRectangle}>
                    <div className={styles.weatherTextBox}>Weather</div>
                </div>
                {!isError ? (
                    <div className={styles.content}>
                        <div className={styles.currentWeatherBox}>
                            <div className={styles.iconContainer}>
                                <Icon icon={icon} />
                            </div>
                            <div className={styles.description}>
                                <div className={styles.weatherNZ}>
                                    {main} | {city}, {country}
                                </div>
                                <div className={styles.temp}>{temp}°C</div>
                                <div className={styles.tempBounds}>
                                    L:{minTemp}°C | H:{maxTemp}°C
                                </div>
                            </div>
                        </div>
                        <div className={styles.hourlyForecast}>{formatHourCards(hourlyData)}</div>
                        <div className={styles.dailyForecast}>{formatDayCards(dailyData)}</div>
                    </div>
                ) : (
                    <div className={styles.errorContainer}>Can not retrieve weather</div>
                )}
            </div>
        </div>
    );
}

export default Weather;
