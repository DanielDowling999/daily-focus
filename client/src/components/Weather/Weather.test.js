import React from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";
import styles from "./style.module.scss";

import Weather from "./";
import dailyData from "./";
//const getTodaysDate = require("../utils/date-helper");
//let todaysDate = getTodaysDate();
//import getWeatherData from "./";
let component;
const lat = -36.85;
const lon = 174.76;

beforeEach(() => {
    component = shallow(<Weather />);
});

test("it renders correctly", () => {
    const shallowRenderer = new ShallowRenderer();
    const snapshotComponent = shallowRenderer.render(<Weather />);
    expect(snapshotComponent).toMatchSnapshot();
});
//All of these cases are almost definitely covered by the snapshot test.
test("it contains DayCards and HourCards", () => {
    expect(component.find(".hourlyForecast")).toHaveLength(1);
    expect(component.find(".dailyForecast")).toHaveLength(1);
});

test("it contains an icon for main", () => {
    expect(component.find(".iconContainer")).toHaveLength(1);
});
