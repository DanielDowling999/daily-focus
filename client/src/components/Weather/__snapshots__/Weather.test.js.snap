// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="weatherWidget"
>
  <div
    className="container"
  >
    <div
      className="headerRectangle"
    >
      <div
        className="weatherTextBox"
      >
        Weather
      </div>
    </div>
    <div
      className="content"
    >
      <div
        className="currentWeatherBox"
      >
        <div
          className="iconContainer"
        >
          <Icon
            icon={null}
          />
        </div>
        <div
          className="description"
        >
          <div
            className="weatherNZ"
          >
             | 
            , 
          </div>
          <div
            className="temp"
          >
            °C
          </div>
          <div
            className="tempBounds"
          >
            L:
            °C | H:
            °C
          </div>
        </div>
      </div>
      <div
        className="hourlyForecast"
      />
      <div
        className="dailyForecast"
      />
    </div>
  </div>
</div>
`;
